cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql

```diff
@@ -97,7 +97,15 @@ class UncontrolledArithConfiguration extends TaintTracking::Configuration {
 
   override predicate isSink(DataFlow::Node sink) { missingGuard(sink.asExpr(), _) }
 
-  override predicate isSanitizer(DataFlow::Node barrier) { bounded(barrier.asExpr()) }
+  override predicate isSanitizer(DataFlow::Node node) {
+    bounded(node.asExpr())
+    or
+    // If this expression is part of bitwise 'and' or 'or' operation it's likely that the value is
+    // only used as a bit pattern.
+    node.asExpr() =
+      any(BinaryBitwiseOperation op | op instanceof BitwiseOrExpr or op instanceof BitwiseAndExpr)
+          .getAnOperand*()
```

Comment:
Why not any `BinaryBitwiseOperation`?  I suppose I'm less confident of `LShiftExpr` but the others all imply the number is used as a bit pattern and/or constrain its range.

Reply:
Is it important that the operands are the sanitizers, not the operation itself?

Reply:
> Is it important that the operands are the sanitizers, not the operation itself?

Yes - the motivating result on ChakraCore is
```
uint64 resplusone_ui64 = ((seed0 + seed1) & mMant) | mExp;
```
where the overflow happens at the addition before it's used in a bitwise operation. (I am skeptical about using `+` rather than `^` there, because of carries, but I don't think that's a problem this query should be looking for)

Reply:
I see, the `&` is telling us that the result of `+` is likely a bit pattern so we aren't scared of it overflowing.

(I suspect the choice of `+` or `^` doesn't matter in this case, as long as there is good enough mixing; i.e. for any fixed `seed0` we can obtain every possibly result (equally) by varying `seed1`, and the same for fixed `seed1`)

Reply:
I completely agree @geoffw0. The problem is that SAMATE defines this `RAND32` macro:
```cpp
/* rand only returns 15 bits, so we xor 3 calls together to get the full result (13 bits overflow, but that is okay) */
// shifting signed values might overflow and be undefined
#define URAND31() (((unsigned)rand()<<30) ^ ((unsigned)rand()<<15) ^ rand())
// choose to produce a positive or a negative number. Note: conditional only evaluates one URAND31
#define RAND32() ((int)(rand() & 1 ? URAND31() : -URAND31() - 1))
```
and that one is used in all of the CWE190 tests. So if we treat `BitwiseXorExpr` as a sanitizer then we lose _all_ of the CWE190 results.

Reply:
OK, that's a good explanation for wanting to leave `BitwiseXorExpr` alone.  Thanks.